AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -0,0 +1,276 @@
+    } else {
+      ((AVPlayerLayer *)_playerNode.layer).player = [[AVPlayer alloc] initWithPlayerItem:_currentItem];
+    }
+
+  }
+  if (_shouldAutoPlay) {
+    [self play];
+  }
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  
+  {
+    ASDN::MutexLocker l(_lock);
+    ((AVPlayerLayer *)_playerNode.layer).player = nil;
+    _shouldBePlaying = NO;
+  }
+}
+
+- (void)setPlayButton:(ASButtonNode *)playButton
+{
+  ASDN::MutexLocker l(_lock);
+  
+  _playButton = playButton;
+  
+  [self addSubnode:playButton];
+  
+  [_playButton addTarget:self action:@selector(play) forControlEvents:ASControlNodeEventTouchUpInside];
+}
+
+- (ASButtonNode *)playButton
+{
+  ASDN::MutexLocker l(_lock);
+  
+  return _playButton;
+}
+
+- (void)setAsset:(AVAsset *)asset
+{
+  ASDN::MutexLocker l(_lock);
+  
+  if (ASObjectIsEqual(((AVURLAsset *)asset).URL, ((AVURLAsset *)_asset).URL)) {
+    return;
+  }
+  
+  _asset = asset;
+  
+  if (self.interfaceState & ASInterfaceStateFetchData) {
+    [self fetchData];
+  }
+}
+
+- (AVAsset *)asset
+{
+  ASDN::MutexLocker l(_lock);
+  return _asset;
+}
+
+- (void)setGravity:(ASVideoGravity)gravity {
+  ASDN::MutexLocker l(_lock);
+  
+  switch (gravity) {
+    case ASVideoGravityResize:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResize;
+      break;
+    case ASVideoGravityResizeAspect:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResizeAspect;
+      break;
+    case ASVideoGravityResizeAspectFill:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResizeAspectFill;
+      break;
+    default:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResizeAspect;
+      break;
+  }
+}
+
+- (ASVideoGravity)gravity;
+{
+  ASDN::MutexLocker l(_lock);
+  
+  if ([((AVPlayerLayer *)_playerNode.layer).contentsGravity isEqualToString:AVLayerVideoGravityResize]) {
+    return ASVideoGravityResize;
+  }
+  if ([((AVPlayerLayer *)_playerNode.layer).contentsGravity isEqualToString:AVLayerVideoGravityResizeAspectFill]) {
+    return ASVideoGravityResizeAspectFill;
+  }
+  
+  return ASVideoGravityResizeAspect;
+}
+
+- (void)play;
+{
+  ASDN::MutexLocker l(_lock);
+  
+  [[((AVPlayerLayer *)_playerNode.layer) player] play];
+  _shouldBePlaying = YES;
+  _playButton.alpha = 0.0;
+  if ([self ready] && ![self.subnodes containsObject:_spinner]) {
+    _spinner = [[ASDisplayNode alloc] initWithViewBlock:^UIView *{
+      UIActivityIndicatorView *spinnnerView = [[UIActivityIndicatorView alloc] initWithFrame:_playButton.frame];
+      spinnnerView.color = [UIColor whiteColor];
+      [spinnnerView startAnimating];
+      
+      return spinnnerView;
+    }];
+    
+    [self addSubnode:_spinner];
+  }
+}
+
+- (BOOL)ready;
+{
+  return [((AVPlayerLayer *)_playerNode.layer) player].currentItem.status != AVPlayerItemStatusReadyToPlay;
```

Comment:
probably worth creating a - (AVPlayer *)_player method so you can just call [self _player] in these places.